Cannot Cast DummyRepository to GitRepository on tasks

Hi, we have a scenario where we can't clone a git repository, so we download the archived sources and run some of our build tasks, which don't involve any versioning, thought the cV task is called (but it's value is not used).

We tried to upgrade from version `1.3.3` to the latest `1.7.0` but it seems that the behaviour of the plugin has changed, from logging a warning message, to failing the build with a `ClassCastException`.

I am aware of https://github.com/allegro/axion-release-plugin/issues/133, but i think this is a different case, as we are not trying to do a release, rather just executing unrelated tasks. 

We would have expected the same behaviour on all the versions. So far we don't require any of the changes on the new versions, but unfortunately this will prevent us from upgrading to any new versions if the behaviour is kept as it currently stands.   

Here is the output for the different versions I've tried, and you can see it started failing from `1.5.0`.

For `1.3.3` version the behaviour is:
```
$ ./gradlew cV --refresh-dependencies
Failed top open repository, trying to work without it org.eclipse.jgit.errors.RepositoryNotFoundException: repository not found: ~/workspace/no-git
Project version: -SNAPSHOT
BUILD SUCCESSFUL
```

For `1.4.1`:
```
$ ./gradlew cV --refresh-dependencies
Failed to open repository, trying to work without it org.eclipse.jgit.errors.RepositoryNotFoundException: repository not found: ~/workspace/no-git
Could not resolve current position on uninitialized repository, returning default

Project version: -SNAPSHOT

BUILD SUCCESSFUL
```

For `1.5.0`, `1.6.0`, `1.7.0`:
```
$ ./gradlew cV --refresh-dependencies
Failed to open repository, trying to work without it org.eclipse.jgit.errors.RepositoryNotFoundException: repository not found: ~/workspace/no-git

FAILURE: Build failed with an exception.

* Where:
Build file '~/workspace/no-git/myProject/build.gradle' line: 15

* What went wrong:
A problem occurred evaluating project ':myProject'.
> Cannot cast object 'pl.allegro.tech.build.axion.release.infrastructure.DummyRepository@345ade19' with class 'pl.allegro.tech.build.axion.release.infrastructure.DummyRepository' to class 'pl.allegro.tech.build.axion.release.infrastructure.git.GitRepository'

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
```

--- a/src/integration/groovy/pl/allegro/tech/build/axion/release/RepositorylessIntegrationTest.groovy
+++ b/src/integration/groovy/pl/allegro/tech/build/axion/release/RepositorylessIntegrationTest.groovy
@@ -1,0 +1,41 @@
+package pl.allegro.tech.build.axion.release
+
+import org.gradle.testkit.runner.GradleRunner
+import org.gradle.testkit.runner.TaskOutcome
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+import spock.lang.Specification
+
+class RepositorylessIntegrationTest extends Specification {
+
+    @Rule
+    TemporaryFolder temporaryFolder = new TemporaryFolder()
+
+    File directory
+
+    void setup() {
+        directory = temporaryFolder.root
+    }
+
+    def "should not fail build when calling currentVersion task on project without repo"() {
+        given:
+        File build = temporaryFolder.newFile('build.gradle')
+        build << """
+        plugins {
+            id 'pl.allegro.tech.build.axion-release'
+        }
+
+        project.version = scmVersion.version
+        """
+
+        when:
+        def result = GradleRunner.create()
+            .withProjectDir(directory)
+            .withPluginClasspath()
+            .withArguments('currentVersion', '-s')
+            .build()
+
+        then:
+        result.task(":currentVersion").outcome == TaskOutcome.SUCCESS
+    }
+}

--- a/src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy
+++ b/src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy
@@ -48,13 +48,11 @@ class VersionResolver {
     private Map readVersions(VersionFactory versionFactory,
                              TagProperties tagProperties,
                              NextVersionProperties nextVersionProperties) {
-        GitRepository repository = (GitRepository) this.repository
         Pattern releaseTagPattern = ~/^${tagProperties.prefix}.*/
         Pattern nextVersionTagPattern = ~/.*${nextVersionProperties.suffix}$/
 
-        Map currentVersionInfo = null
-        Map previousVersionInfo = null
-        List<TagsOnCommit> allTaggedCommits = repository.firstTaggedCommitAsList(releaseTagPattern)
+        Map currentVersionInfo, previousVersionInfo
+        List<TagsOnCommit> allTaggedCommits = [repository.latestTags(releaseTagPattern)]
         currentVersionInfo = versionFromTaggedCommits(allTaggedCommits, false, nextVersionTagPattern, versionFactory)
 
         TagsOnCommit previousTags = repository.latestTags(releaseTagPattern)
@@ -77,13 +75,10 @@ class VersionResolver {
     private Map readVersionsByHighestVersion(VersionFactory versionFactory,
                                               TagProperties tagProperties,
                                              NextVersionProperties nextVersionProperties) {
-        GitRepository repository = (GitRepository) this.repository
         Pattern releaseTagPattern = ~/^${tagProperties.prefix}.*/
         Pattern nextVersionTagPattern = ~/.*${nextVersionProperties.suffix}$/
 
-        Map currentVersionInfo = null
-        Map previousVersionInfo = null
-
+        Map currentVersionInfo, previousVersionInfo
         List<TagsOnCommit> allTaggedCommits = repository.taggedCommits(releaseTagPattern)
 
         currentVersionInfo = versionFromTaggedCommits(allTaggedCommits, false, nextVersionTagPattern, versionFactory)

--- a/src/main/groovy/pl/allegro/tech/build/axion/release/domain/scm/ScmRepository.groovy
+++ b/src/main/groovy/pl/allegro/tech/build/axion/release/domain/scm/ScmRepository.groovy
@@ -20,6 +20,8 @@ interface ScmRepository {
 
     TagsOnCommit latestTags(Pattern pattern, String sinceCommit)
 
+    List<TagsOnCommit> taggedCommits(Pattern pattern)
+
     boolean remoteAttached(String remoteName);
 
     boolean checkUncommittedChanges()

--- a/src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/DryRepository.groovy
+++ b/src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/DryRepository.groovy
@@ -65,6 +65,11 @@ class DryRepository implements ScmRepository {
     }
 
     @Override
+    List<TagsOnCommit> taggedCommits(Pattern pattern) {
+        return delegateRepository.taggedCommits(pattern)
+    }
+
+    @Override
     boolean remoteAttached(String remoteName) {
         return true
     }

--- a/src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/DummyRepository.groovy
+++ b/src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/DummyRepository.groovy
@@ -62,6 +62,12 @@ class DummyRepository implements ScmRepository {
         return new TagsOnCommit(null, [], false)
     }
 
+    @Override
+    List<TagsOnCommit> taggedCommits(Pattern pattern) {
+        logger.quiet("Could not resolve current position on uninitialized repository, returning default")
+        return [new TagsOnCommit(null, [], false)]
+    }
+
 
     @Override
     boolean remoteAttached(String remoteName) {

--- a/src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.groovy
+++ b/src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.groovy
@@ -169,10 +169,12 @@ class GitRepository implements ScmRepository {
         )
     }
 
+    @Override
     TagsOnCommit latestTags(Pattern pattern) {
         return latestTagsInternal(pattern, null, true)
     }
 
+    @Override
     TagsOnCommit latestTags(Pattern pattern, String sinceCommit) {
         return latestTagsInternal(pattern, sinceCommit, false)
     }
@@ -182,10 +184,7 @@ class GitRepository implements ScmRepository {
         return taggedCommits.isEmpty() ? TagsOnCommit.empty() : taggedCommits[0]
     }
 
-    List<TagsOnCommit> firstTaggedCommitAsList(Pattern pattern) {
-        return taggedCommitsInternal(pattern, null, true, true)
-    }
-
+    @Override
     List<TagsOnCommit> taggedCommits(Pattern pattern) {
         return taggedCommitsInternal(pattern, null, true, false)
     }

